ddtrace/settings/config.py

```diff
@@ -141,6 +159,229 @@ def get_error_ranges(error_range_str):
+    def __init__(self, *args, **kwargs):
+        # type: (*Any, **Any) -> None
+        pass
+
+    def get(self):
+        # type: () -> _ConfigResult
+        return {"value": None, "raw": None, "source": "env_var"}
+
+
+class _ConfigSourceEnvMulti(_ConfigSourceEnv):
+    def __init__(self, *envs, **kwargs):
+        # type: (*_ConfigSourceEnv, **Any) -> None
+        resolver = typing.cast("Optional[Callable[[_ConfigSourceEnv], _ConfigResult]]", kwargs.get("resolver"))
+        self.envs = envs
+        self.resolver = resolver or self._default_resolver
+
+    @staticmethod
+    def _default_resolver(*envs):
+        # type: (_ConfigSourceEnv) -> _ConfigResult
+        """
+        Default is to take the first env var that has a value.
+        """
+        for e in envs:
+            v = e.get()
+            if v["value"] is not None:
+                return v
+        return {
+            "value": None,
+            "raw": None,
+            "source": "env_var",
+        }
+
+    def get(self):
+        # type: () -> _ConfigResult
+        return self.resolver(*self.envs)
+
+
+class _ConfigSourceProgrammatic(_ConfigSource):
+    """A configuration source that is set programmatically by the user."""
+
+    def __init__(
+        self,
+        factory=None,  # type: Optional[Callable[[Any], Any]]
+    ):
+        # type: (...) -> None
+        self._value = None
+        self._raw = None
+        self._factory = factory
+
+    def set(self, val):
+        # type: (Any) -> None
+        self._raw = val
+        self._value = val
+        if val is not None and self._factory is not None:
+            self._value = self._factory(val)
+
+    def get(self):
+        # type: () -> _ConfigResult
+        return {
+            "value": self._value,
+            "raw": self._raw,
+            "source": "code",
+        }
+
+
+class _ConfigItem(object):
+    """A configuration item that can be queried for a value.
+
+    By default, the value is resolved in the following order:
+        1. programmatic
+        2. environ
+        3. default
+    """
+
+    _sentinel = object()
+
+    def __init__(
+        self,
+        key,  # type: str
+        type_,  # type: str
+        default,  # type: Any
+        environ=None,  # type: Optional[_ConfigSourceEnv]
+        programmatic=None,  # type: Optional[_ConfigSourceProgrammatic]
+        resolve=None,  # type: Optional[Callable[[Any, _ConfigSourceEnv, _ConfigSourceProgrammatic], _ConfigResult]]
+        metadata=None,  # type: Optional[Dict[str, Any]]
+    ):
+        # type: (...) -> None
+        self.key = key
+        self.type = type_
+        self._default_factory = default if callable(default) else lambda: default
+        self.environ = environ or _ConfigSourceEnvNone()  # type: _ConfigSourceEnv
+        self.programmatic = programmatic or _ConfigSourceProgrammatic()  # type: _ConfigSourceProgrammatic
+        self.resolve = self._default_resolve if resolve is None else resolve
```

Comment:
Is overriding the resolver necessary? It seems like overkill. We can probably define one resolution order for all configurations (ex: rc -> code -> envar ->  default)